Source/items.cpp

```diff
@@ -1171,27 +1171,61 @@ void GetStaffPower(Item &item, int lvl, int bs, bool onlygood)
-			strncpy(item._iIName, fmt::format(/* TRANSLATORS: Constructs item names. Format: {Prefix} {Item}. Example: King's Long Sword */ _("{0} {1}"), _(ItemPrefixes[preidx].PLName), item._iIName).c_str(), sizeof(item._iIName) - 1);
-			item._iIName[sizeof(item._iIName) - 1] = '\0';
 			item._iMagical = ITEM_QUALITY_MAGIC;
 			SaveItemAffix(item, ItemPrefixes[preidx]);
 			item._iPrePower = ItemPrefixes[preidx].power.type;
 		}
 	}
-	if (!StringInPanel(item._iIName)) {
-		strcpy(item._iIName, _(AllItemsList[item.IDidx].iSName));
-		if (preidx != -1) {
-			strncpy(item._iIName, fmt::format(/* TRANSLATORS: Constructs item names. Format: {Prefix} {Item}. Example: King's Sword */ _("{0} {1}"), _(ItemPrefixes[preidx].PLName), item._iIName).c_str(), sizeof(item._iIName) - 1);
+
+	const char *baseName = _(AllItemsList[item.IDidx].iName);
+	const char *shortName = _(AllItemsList[item.IDidx].iSName);
+	const char *spellName = pgettext("spell", spelldata[bs].sNameText);
+	const char *normalFmt = pgettext("spell", /* TRANSLATORS: Constructs item names. Format: {Item} of {Spell}. Example: War Staff of Firewall */ "{0} of {1}");
+
+	strncpy(item._iName, fmt::format(normalFmt, baseName, spellName).c_str(), sizeof(item._iName) - 1);
+	item._iName[sizeof(item._iName) - 1] = '\0';
+	if (!StringInPanel(item._iName)) {
+		strncpy(item._iName, fmt::format(normalFmt, shortName, spellName).c_str(), sizeof(item._iName) - 1);
+		item._iName[sizeof(item._iName) - 1] = '\0';
+	}
+
+	if (preidx != -1) {
+		const char *magicFmt = pgettext("spell", /* TRANSLATORS: Constructs item names. Format: {Prefix} {Item} of {Spell}. Example: King's War Staff of Firewall */ "{0} {1} of {2}");
+		const char *prefixName = _(ItemPrefixes[preidx].PLName);
+		strncpy(item._iIName, fmt::format(magicFmt, prefixName, baseName, spellName).c_str(), sizeof(item._iIName) - 1);
+		item._iIName[sizeof(item._iIName) - 1] = '\0';
+		if (!StringInPanel(item._iIName)) {
+			strncpy(item._iIName, fmt::format(magicFmt, prefixName, shortName, spellName).c_str(), sizeof(item._iIName) - 1);
 			item._iIName[sizeof(item._iIName) - 1] = '\0';
 		}
-		strncpy(item._iIName, fmt::format(_(/* TRANSLATORS: Constructs item names. Format: {Prefix Item} of {Spell}. Example: King's Staff of Firewall */ "{0} of {1}"), item._iIName, pgettext("spell", spelldata[bs].sNameText)).c_str(), sizeof(item._iIName) - 1);
-		item._iIName[sizeof(item._iIName) - 1] = '\0';
-		if (item._iMagical == ITEM_QUALITY_NORMAL)
-			strcpy(item._iName, item._iIName);
+	} else {
+		strncpy(item._iIName, item._iName, sizeof(item._iIName) - 1);
+		item._iName[sizeof(item._iName) - 1] = '\0';
 	}
+
 	CalcItemValue(item);
 }
 
+namespace {
+
+std::string GenerateMagicItemName(const string_view &baseNamel, int preidx, int sufidx)
+{
+	if (preidx != -1 && sufidx != -1) {
+		const char *fmt = _(/* TRANSLATORS: Constructs item names. Format: {Prefix} {Item} of {Spell}. Example: King's Long Sword of the Whale */ "{0} {1} of {2}");

```

Comment:
```suggestion
		const char *fmt = _(/* TRANSLATORS: Constructs item names. Format: {Prefix} {Item} of {Suffix}. Example: King's Long Sword of the Whale */ "{0} {1} of {2}");
```
Guessing this was copied from the staff suffix TL note.

Reply:
Thanks :D